Fixes #25224 - Fix RHEL8 arches in RH repo page

johnpmitsch opened the pull request:
RHEL8 paths in the CDN no longer use substitutions, i.e.
$basearch. The paths are now fixed, set paths that looks
like "content/dist/rhel8/8.0/x86_64/baseos/os/"

We make a lot of assumptions on  both the back-end and
front-end that the EL6 and EL7 substitution paths exist.
Because of this, I added methods to grab the info from
the url and add it to the substitutions hash we have
been using.

To review, check that:
- EL5, 6, and 7 repos act as expected on the RH repos page
- You can now view the arch for the RHEL8 repos

johnpmitsch commented:
@jturel @jlsherrill Do we want to grab the release version as well as arch? We normally show this info i.e.

![here](http://www.clipular.com/c/6457330867372032.png?k=RlrZGeyyqme3Ii8TD6UogUvDxFk)

I can grab the release version from the URL, but will have to make an exception for layered reposiitories which don't have a release version.

beav commented on app/lib/katello/util/path_with_substitutions.rb:
```diff
@@ -12,12 +12,35 @@ def initialize(path, substitutions)
+
+      def rhel_eight_arch
+        # arch for RHEL8 repo paths is always the in the 5th position of the url i.e.
+        # /content/dist/rhel8/8.0/x86_64/baseos/os/ OR
```
looks like you are missing the 2nd half of the `OR`

johnpmitsch replied:
bad copy paste thanks!

beav commented on app/lib/katello/util/path_with_substitutions.rb:
```diff
@@ -12,12 +12,35 @@ def initialize(path, substitutions)
         @substitutions = substitutions
         @path = path
         @resolved = []
+        create_rhel_eight_substitutions
```
could this be renamed to something like `create_substitutions`? If the release-specific magic could live in its own class where we just add special cases as needed, that would be good.

johnpmitsch replied:
yes, I was thinking of a better way to structure this, I was thinking of a helper module with the methods that could be included here, but I'm not sure I can change the initializer easily. 

Maybe we could call RHEL8 and higher something? Like `modern_rhel` or something like that, so this would be `create_modern_rhel_substiutions`.

jlsherrill replied:
I think the term 'substitutable' or 'non_substitutable' would be better, and a method named 'extract_hardcoded_substitutions' or something like that would be just as clear?

jlsherrill commented on app/lib/katello/util/path_with_substitutions.rb:
```diff
@@ -12,12 +12,35 @@ def initialize(path, substitutions)
+      def rhel_eight_arch
+        # arch for RHEL8 repo paths is always the in the 5th position of the url i.e.
+        # /content/dist/rhel8/8.0/x86_64/baseos/os/ OR
+        split_path[4]
```
This seems like it would be extremely fragile and could easily break.  Would it make more sense to have a list of arches and look for any of those, i feel like new arches will be introduced few times than the cdn will introduce some new path structure.

johnpmitsch commented:
@beav @jlsherrill I pushed up [a new approach](https://github.com/Katello/katello/pull/7882/commits/90d54a366b0ffc17754ffd6dcbed829af35af33d) that uses a list of arches to check for in the path when there are no substitutions provided. I admit it is not without flaws, but it uncouples the logic from checking RHEL8.

beav commented on app/lib/katello/util/path_with_substitutions.rb:
```diff
@@ -12,12 +14,23 @@ def initialize(path, substitutions)
 
+      def generate_substitutions_from_path
+        arches = split_path & ARCHITECTURES
+        arch = arches.first
```
should this raise an exception if arches is bigger than 1?

johnpmitsch replied:
I'm not sure.. that would mean the repo would have a path like `/content/dist/rhel8/8/x86_64/product/x86_64` so to generate an incorrect substitution, it would have to have two arches in its path and one be incorrect, i.e. `/content/dist/rhel8/8/aarch64/product/x86_64`

One thing I was thinking of adding is a check in the fifth position for the arch first since that seems to be the agreed upon convention for RHEL8.

That would look like this:
- Check 5th position of the path for arch in ARCHITECTURES, if exists, use that in substitutions and skip checking the full path
- Check full path for arch in ARCHICTECTURES, if any exist, use the first one found in substitutions.
- If no arches found, do nothing.

This would at least check the expected place for the arch first before checking the whole path

johnpmitsch commented:
@beav @jlsherrill  I updated the approach to what I discussed. I'm not sure if there is a better way than getting info from the path if no substitutions are provided, but I am open to suggestions!

jlsherrill commented on app/lib/katello/util/path_with_substitutions.rb:
```diff
@@ -12,12 +14,28 @@ def initialize(path, substitutions)
+      def generate_substitutions_from_path
+        # in RHEL8, arch is typically in the 5th position of the path, so we check here first
+        if split_path.length >= 5 && ARCHITECTURES.include?(split_path[4])
+          @substitutions["basearch"] = split_path[4]
```
I'm not sure i understand the value this adds, and i'm not sure i agree that it was agreed upon as being the '5th' directory in the path for rhel8, as i've already found a repo that conflicts with that rule.

johnpmitsch replied:
@jlsherrill would the repo you found (or exceptions you are thinking of) not be covered by the else block? I think I covered this and I even added a test with a repo path that didn't have the arch in the fifth postion.

I discussed the reasoning a bit here https://github.com/Katello/katello/pull/7882#discussion_r241104554 but the basic thought is let's look where RHEL8 is most likely to put the arch first and then check the whole path if its not there.

jlsherrill replied:
It would, i just don't understand the value that the first check provides over the 2nd check.  The only condition the first check would catch is if there happens to be another arch in the first 4 portions of the path, but we want this particular one?  That seems extremely unlikely to me and a very specific scenario.

johnpmitsch replied:
@jlsherrill I think its worthy to check the expected place for the arch first if we are hearing that they CDN team is trying to place the arch there for RHEL8 repos, even if it is not a guarantee. I understand your point that it would be a unique edge case to have an arch come before, but I also don't see the harm in being a bit more specific about where to check first? I'm just going off what was shared with us

jlsherrill replied:
I just don't see any value in it?  Especially if RCM isn't even honoring it already.

If i was working in this code, the first thing i'd do is look at this branch condition and wonder why we didn't just use the else statement, since that seems to cover this condition.  Then i'd read about the rhel8 statement and worry that if i touch it i might break something, but never really know what or why, which brings me back to the fact i don't see any value in it (unless I am missing something?)

johnpmitsch replied:
I'm not sure we are going to come to an agreement :) I think we should try to look where the RCM team is making an effort to put the arch first, even if there have been a small number of cases where they haven't put the arch there, the overwhelming majority of RHEL8 repos have it there and it been stated that the arch intended to go in a consistent location.

I think we need some outside input - @beav and/or @jturel do you mind weighing in with your opinion(s)?

jlsherrill replied:
yep, sounds good :)

jturel replied:
Joining late but I think I'm following and I understand the question here to be: 'Do we need the 5th position checking for arch in this method' and my response is:

If the tests pass right now (given both branches of `generate_subs_from_path`) and continue to pass without the aforementioned check then we don't need to look at 5th position explicitly. If test fails and it's a valid behavior, then we need to have the check the way it's written or some other way

parthaa replied:
My 2 cents with limited knowledge.. Unless I am missing something the "else" condition handles RHEL8 anyway right? Don't see the need for a make a specific check. If you are worried about paths having 2 arches in it that's probably an RCM issue.

jlsherrill commented on app/lib/katello/util/path_with_substitutions.rb:
```diff
@@ -1,6 +1,8 @@
 module Katello
   module Util
     class PathWithSubstitutions
+      ARCHITECTURES = ["x86_64", "s390x", "ppc64le", "aarch64" ].freeze
```
I would add 'multiarch' and 'ppc64' here as well to cover the container and rhel7 cases

jlsherrill commented on app/lib/katello/util/path_with_substitutions.rb:
```diff
@@ -12,12 +14,28 @@ def initialize(path, substitutions)
+      end
+
+      def split_path
+        @split ||= @path.split('/').reject(&:blank?)
```
I might would lowercase this as well, just in case some upper case letters are used somewhere in the manifest

jlsherrill commented:
[test katello]

jlsherrill commented:
This looks like a legit failure:
```
unexpected invocation: #<Mock:0xae17440>.get("/content/dist/x86_64/atomic/7/7Server/x86_64/kickstart/repodata/repomd.xml")
satisfied expectations:
- allowed any number of times, not yet invoked: Katello::Ping.ping(any_parameters)
- allowed any number of times, not yet invoked: Cert::Certs.ssl_client_key(any_parameters)
- allowed any number of times, not yet invoked: Cert::Certs.ssl_client_cert(any_parameters)
- allowed any number of times, not yet invoked: Cert::Certs.ca_cert(any_parameters)
(Minitest::Assertion)
/var/lib/workspace/workspace/katello-pr-test/test/lib/util/cdn_var_subsitutor_test.rb:79
```

johnpmitsch commented:
@jlsherrill I updated the test,  If there are no substitutions, we now check the URL for an arch and use that, so passing a URL with `$basearch` and `x86_64` in the URL will resolve correctly. I just changed the arch to `fakearch`.

Another option we could do is only check the URL for the arch if there are no substitutions AND no placeholders like `$basearch`

jlsherrill commented:
```
Another option we could do is only check the URL for the arch if there are no substitutions AND no placeholders like $basearch
``

I think this is a better solution, if the path has $basearch in it, we shouldn't be trying to guess the arch, as the user HAS to provide it.

johnpmitsch commented:
@jlsherrill updated! The previously failing test is now passing (without modification)

jlsherrill commented:
[test katello]

johnpmitsch commented:
[test katello]

jlsherrill commented on app/lib/katello/util/path_with_substitutions.rb:
```diff
@@ -14,7 +14,7 @@ def initialize(path, substitutions)
         @path = path
         @resolved = []
-        generate_substitutions_from_path if @substitutions.empty?
+        generate_substitutions_from_path if @substitutions.empty? && substitutions_needed.empty?
```
How would this handle if you had a url like:   /content/$releasever/x86_64/,  wouldn't that fail to extract the substitutions?   Would it make more sense to check for the presence  of $basearch  specifically?

johnpmitsch replied:
The new logic means it tries to generate substitutions when it both a) does not have any substitution variables in the name and b) is not passed with any substitutions to fill variables, so `/content/$releasever/x86_64/` would either need a substitution passed with it or it would error like normally.

What would you want to check regarding `$releasever`? This should be handled as normal with RHEL 7 and under paths

jlsherrill replied:
I think, if i'm reading this correct, if a url such as ` /content/$releasever/x86_64/` was used, substitutions_needed wouldn't be empty, so `generate_substitutions_from_path` wouldn't get called, right?

johnpmitsch commented:
[test katello]

johnpmitsch commented:
@jlsherrill updated

johnpmitsch commented:
[test katello]

johnpmitsch commented:
[test katello]

johnpmitsch commented:
[test katello]

johnpmitsch commented:
[test katello]

johnpmitsch commented:
[test katello]

jlsherrill commented:
[test katello]

beav commented:
no further comment from me, thanks @johnpmitsch !
